controlplane/kubeadm/internal/controllers/status.go

```diff
@@ -135,3 +142,257 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, contro
+		return
+	}
+
+	v1beta2conditions.Set(kcp, *readyCondition)
+}
+
+func setMachinesUpToDateCondition(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, machines collections.Machines) {
+	if len(machines) == 0 {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition,
+			Status: metav1.ConditionTrue,
+			Reason: controlplanev1.KubeadmControlPlaneMachinesUpToDateNoReplicasV1Beta2Reason,
+		})
+		return
+	}
+
+	upToDateCondition, err := v1beta2conditions.NewAggregateCondition(
+		machines.UnsortedList(), clusterv1.MachineUpToDateV1Beta2Condition,
+		v1beta2conditions.TargetConditionType(controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition),
+	)
+	if err != nil {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  controlplanev1.KubeadmControlPlaneMachinesUpToDateInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+
+		log := ctrl.LoggerFrom(ctx)
+		log.Error(err, fmt.Sprintf("Failed to compute %s condition", controlplanev1.KubeadmControlPlaneMachinesUpToDateV1Beta2Condition))
+		return
+	}
+
+	v1beta2conditions.Set(kcp, *upToDateCondition)
+}
+
+func calculateMissingReferencesMessage(kcp *controlplanev1.KubeadmControlPlane, infraMachineTemplateNotFound bool) string {
+	if infraMachineTemplateNotFound {
+		return fmt.Sprintf("because %s does not exist", kcp.Spec.MachineTemplate.InfrastructureRef.Kind)
+	}
+	return ""
+}
+
+func setRemediatingCondition(ctx context.Context, kcp *controlplanev1.KubeadmControlPlane, unhealthyMachines collections.Machines) {
+	if len(unhealthyMachines) == 0 {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:   controlplanev1.KubeadmControlPlaneRemediatingV1Beta2Condition,
+			Status: metav1.ConditionFalse,
+			Reason: controlplanev1.KubeadmControlPlaneNotRemediatingV1Beta2Reason,
+		})
+		return
+	}
+
+	message := fmt.Sprintf("%d machines not healthy", len(unhealthyMachines))
+
+	// TODO: Bring together externally remediated machines and owner remediated machines
+	remediatingCondition, err := v1beta2conditions.NewAggregateCondition(
+		unhealthyMachines.UnsortedList(), clusterv1.MachineOwnerRemediatedV1Beta2Condition,
+		v1beta2conditions.TargetConditionType(controlplanev1.KubeadmControlPlaneRemediatingV1Beta2Condition),
+	)
+	if err != nil {
+		v1beta2conditions.Set(kcp, metav1.Condition{
+			Type:    controlplanev1.KubeadmControlPlaneRemediatingV1Beta2Condition,
+			Status:  metav1.ConditionUnknown,
+			Reason:  controlplanev1.KubeadmControlPlaneRemediatingInternalErrorV1Beta2Reason,
+			Message: "Please check controller logs for errors",
+		})
+
+		log := ctrl.LoggerFrom(ctx)
+		log.Error(err, fmt.Sprintf("Failed to compute %s condition", controlplanev1.KubeadmControlPlaneRemediatingV1Beta2Condition))
```

Comment:
Maybe same as above for ready & update

(I"m fine with either the KCP or the MS variant or a mixture of both, would be just nice to ~ use the same)